testbed/tests/test_window.py

```diff
@@ -479,6 +479,60 @@ async def test_full_screen(second_window, second_window_probe):
         assert not second_window_probe.is_full_screen
         assert second_window_probe.content_size == initial_content_size
 
+    @pytest.mark.parametrize(
+        "second_window_kwargs",
+        [dict(title="Secondary Window", position=(200, 150))],
+    )
+    async def test_on_gain_focus(
+        main_window, main_window_probe, second_window, second_window_probe
+    ):
+        on_gain_focus_handler = Mock()
+        main_window.on_gain_focus = on_gain_focus_handler
+        main_window.app.current_window = second_window
+        await second_window_probe.wait_for_window(
+            "Setting second window as current window"
+        )
+        main_window.app.current_window = main_window
+        await main_window_probe.wait_for_window("Setting main window as current window")
+        on_gain_focus_handler.assert_called_once_with(main_window)
+
+    @pytest.mark.parametrize(
+        "second_window_kwargs",
+        [dict(title="Secondary Window", position=(200, 150))],
+    )
+    async def test_on_lose_focus(
+        main_window, main_window_probe, second_window, second_window_probe
+    ):
+        on_lose_focus_handler = Mock()
+        main_window.on_lose_focus = on_lose_focus_handler
+        main_window.app.current_window = main_window
+        await main_window_probe.wait_for_window("Setting main window as current window")
+        second_window.app.current_window = second_window
+        await second_window_probe.wait_for_window(
+            "Setting second window as current window"
+        )
+        on_lose_focus_handler.assert_called_once_with(main_window)
+        main_window.app.current_window = main_window
+        await main_window_probe.wait_for_window("Setting main window as current window")
+
+    async def test_on_show(main_window, main_window_probe):
```

Comment:
As with the focus tests - it's easier to wrap show/hide into a single test, because you're going to have a hide a shown window (and vice versa) for a full test of this functionality.

Reply:
Done. Thanks!